layers_from_onnx/tune_eval_vm.py

```diff
@@ -0,0 +1,673 @@
+import os
+import numpy as np
+
+import tvm
+import tvm.testing
+from tvm import autotvm
+from tvm import relay
+from tvm.contrib import utils, ndk
+from tvm.runtime.vm import VirtualMachine
+
+import argparse
+
+from tvm.relay.op import register_mixed_precision_conversion
+
+# TODO(amalyshe) current ugly solution with global variable should be substituted to
+# more convenient
+conv2d_acc = "float32"
+
+# Pick a priority > 10 to overwrite defaults, higher priorities take precedence
+@register_mixed_precision_conversion("nn.conv2d", level=11)
+def conv2d_mixed_precision_rule(call_node: "relay.Call", mixed_precision_type: str):
+    global conv2d_acc
+    return [
+        # always do main calculation in mixed_precision_type
+        relay.transform.mixed_precision.MIXED_PRECISION_ALWAYS,
+        # the dtype for the accumulator
+        conv2d_acc,
+        # the output dtype for the operation (usually fp16)
+        mixed_precision_type,
+    ]
+
+@register_mixed_precision_conversion("nn.dense", level=11)
+def conv2d_mixed_precision_rule(call_node: "relay.Call", mixed_precision_type: str):
+    global conv2d_acc
+    return [
+        # always do main calculation in mixed_precision_type
+        relay.transform.mixed_precision.MIXED_PRECISION_ALWAYS,
+        # the dtype for the accumulator
+        conv2d_acc,
+        # the output dtype for the operation (usually fp16)
+        mixed_precision_type,
+    ]
+
+def convert_to_dtype(mod, dtype):
+    # downcast to float16
+    if dtype == "float16" or dtype == "float16_acc32":
+        global conv2d_acc
+        conv2d_acc = "float16" if dtype == "float16" else "float32"
+        from tvm.ir import IRModule
+        mod = IRModule.from_expr(mod)
+        seq = tvm.transform.Sequential(
+            [
+                relay.transform.InferType(),
+                relay.transform.ToMixedPrecision()
+            ]
+        )
+        with tvm.transform.PassContext(
+                config={"relay.ToMixedPrecision.keep_orig_output_dtype": True},
+                opt_level=3):
+            mod = seq(mod)
+    return mod
```

Comment:
Move this part to a new module and import it in `evaluate.py` and in this script.